nilearn/_utils/niimg_conversions.py

```diff
@@ -55,41 +61,62 @@ def check_niimg(niimg, ensure_3d=False):
 
     Its application is idempotent.
     """
+    # If it's an iterator, it's a 4d image
     if hasattr(niimg, "__iter__") and not isinstance(niimg, string_types):
-        if ensure_3d:
-            raise TypeError("A 3D image is expected, but an iterable was"
-                " given: %s" % short_repr(niimg))
-        if hasattr(niimg, "__len__") and len(niimg) == 0:
-            raise TypeError('An empty object - %r - was passed instead of an '
-                            'image or a list of images' % niimg)
-        return concat_niimgs(niimg)
-
-    niimg = load_img(niimg)
-
-    if ensure_3d:
-        shape = niimg.shape
-        if len(shape) == 3:
-            pass
-        elif (len(shape) == 4 and shape[3] == 1):
-            # "squeeze" the image.
-            data = _safe_get_data(niimg)
-            affine = niimg.get_affine()
-            niimg = new_img_like(niimg, data[:, :, :, 0], affine)
-        else:
-            raise TypeError("A 3D image is expected, but an image "
-                "with a shape of %s was given." % (shape, ))
+        return check_niimg_4d(niimg)
+
+    # Otherwise, it should be a filename or a SpatialImage, we load it
+    niimg = load_niimg(niimg)
+
+    if atleast_4d and len(niimg.shape) == 3:
+        data = niimg.get_data().view()
+        data.shape = data.shape + (1, )
+        niimg = new_img_like(niimg, data, niimg.get_affine())
+
     return niimg
 
 
-def _to_4d(data):
-    """ Internal function to cast a 3D ndarray to a 4D one by adding a
-        new axis at the end
+def check_niimg_3d(niimg):
+    """Check that niimg is a proper 3D niimg. Turn filenames into objects.
+
+    Parameters
+    ----------
+    niimg: Niimg-like object
+        See http://nilearn.github.io/building_blocks/manipulating_mr_images.html#niimg.
+        If niimg is a string, consider it as a path to Nifti image and
+        call nibabel.load on it. If it is an object, check if get_data()
+        and get_affine() methods are present, raise TypeError otherwise.
+
+    Returns
+    -------
+    result: 3D Niimg-like object
+        Result can be nibabel.Nifti1Image or the input, as-is. It is guaranteed
+        that the returned object has get_data() and get_affine() methods.
+
+    Notes
+    -----
+    In nilearn, special care has been taken to make image manipulation easy.
+    This method is a kind of pre-requisite for any data processing method in
+    nilearn because it checks if data have a correct format and loads them if
+    necessary.
+
+    Its application is idempotent.
```

Comment:
how is `idempotent` not international??

it doesn't mean "identical". It means that if you apply it n times, where n
>= 1, it is like applying it once.

On Thu, Mar 19, 2015 at 8:58 PM, Danilo Bzdok <notifications@github.com>
wrote:

> In nilearn/_utils/niimg_conversions.py
> <https://github.com/nilearn/nilearn/pull/505#discussion_r26791583>:
>
> > +        and get_affine() methods are present, raise TypeError otherwise.
> > +
> > +    Returns
> > +    -------
> > +    result: 3D Niimg-like object
> > +        Result can be nibabel.Nifti1Image or the input, as-is. It is guaranteed
> > +        that the returned object has get_data() and get_affine() methods.
> > +
> > +    Notes
> > +    -----
> > +    In nilearn, special care has been taken to make image manipulation easy.
> > +    This method is a kind of pre-requisite for any data processing method in
> > +    nilearn because it checks if data have a correct format and loads them if
> > +    necessary.
> > +
> > +    Its application is idempotent.
>
> Given the internationality of user pool, perhaps exchange idempotentn by
> something simpler - "identical"
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nilearn/nilearn/pull/505/files#r26791583>.
>